File: backend/src/controllers/resume.ts
import { Response } from 'express'
import { prisma } from '../lib/db'
import type { AuthenticatedRequest } from '../middlewares/auth'
import { calculateReadinessScore } from './profile'
import multer from 'multer'
import path from 'path'
import fs from 'fs'
import os from 'os'

import { uploadFileToFirebase } from '../lib/storage'

// ── Configure multer memory storage ──────────────────────────────────────────
const storage = multer.memoryStorage()

const fileFilter = (_req: any, file: Express.Multer.File, cb: multer.FileFilterCallback) => {
  const allowed = ['.pdf', '.docx', '.doc']
  const ext = path.extname(file.originalname).toLowerCase()
  if (allowed.includes(ext)) {
    cb(null, true)
  } else {
    cb(new Error('Only PDF and DOCX files are accepted'))
  }
}

export const upload = multer({
  storage,
  fileFilter,
  limits: { fileSize: 10 * 1024 * 1024 } // 10MB limit
})

// ── ATS Analysis helper ─────────────────────────────────────────────────────
async function analyzeResume(userId: string, fileName: string, fileBuffer: Buffer) {
  const profile = await prisma.profile.findUnique({
    where: { userId },
    include: { skills: true }
  })

  const userSkills = profile?.skills.map(s => s.name.toLowerCase()) || []

  // Direct text search in binary string of file buffer
  const bufferStr = fileBuffer.toString('binary').toLowerCase()
  const atsKeywords = ['javascript', 'typescript', 'python', 'react', 'node', 'sql', 'git', 'docker', 'ci/cd', 'system design']
  
  // Matched if in user's profile skills OR found in resume buffer text
  const matched = atsKeywords.filter(kw => userSkills.includes(kw) || bufferStr.includes(kw))
  const missing = atsKeywords.filter(kw => !matched.includes(kw))

  const baseScore = 40
  const skillBonus = Math.min(matched.length * 6, 36)
  const profileBonus = (profile?.bio ? 8 : 0) + (profile?.collegeName ? 8 : 0) + (userSkills.length >= 3 ? 8 : 0)
  const atsScore = Math.min(baseScore + skillBonus + profileBonus, 100)

  const keywordReport = [
    `✅ Resume file "${fileName}" successfully uploaded and parsed.`,
    `✅ ${matched.length} technical keywords matched: ${matched.length > 0 ? matched.join(', ') : 'None yet — add skills in your portfolio.'}`,
    atsScore < 70
      ? `⚠️ ATS score is ${atsScore}%. Add more skills and fill your bio to improve visibility.`
      : `✅ ATS score of ${atsScore}% — excellent keyword optimization!`,
    missing.length > 0
      ? `💡 Consider adding these keywords if you have them: "${missing.slice(0, 4).join('", "')}"`
      : `💡 Tip: Add your LinkedIn and GitHub URLs to your resume header for extra recruiter visibility.`
  ]

  return { atsScore, keywordReport }
}

// ── Upload endpoint ─────────────────────────────────────────────────────────
export const uploadResume = async (req: AuthenticatedRequest, res: Response) => {
  if (!req.user) return res.status(401).json({ error: 'Unauthorized' })

  try {
    const file = req.file
    if (!file) {
      return res.status(400).json({ error: 'No file uploaded. Please select a PDF or DOCX file.' })
    }

    // Upload to Firebase Storage (falls back to local filesystem under resumes/ folder)
    const fileUrl = await uploadFileToFirebase(
      file.buffer,
      file.originalname,
      file.mimetype,
      'resumes',
      req.user.id
    )

    const { atsScore, keywordReport } = await analyzeResume(req.user.id, file.originalname, file.buffer)

    const newResume = await prisma.resume.create({
      data: {
        userId: req.user.id,
        fileName: file.originalname,
        fileUrl,
        atsScore,
        keywordReport: JSON.stringify(keywordReport)
      }
    })

    // Update Career Readiness Score
    await calculateReadinessScore(req.user.id)

    res.status(201).json(newResume)
  } catch (error) {
    console.error('Resume upload error:', error)
    res.status(500).json({ error: 'Internal server error evaluating resume' })
  }
}

// ── Get all resumes ─────────────────────────────────────────────────────────
export const getResumes = async (req: AuthenticatedRequest, res: Response) => {
  if (!req.user) return res.status(401).json({ error: 'Unauthorized' })
  try {
    const list = await prisma.resume.findMany({
      where: { userId: req.user.id },
      orderBy: { uploadedAt: 'desc' }
    })
    res.json(list)
  } catch (error) {
    res.status(500).json({ error: 'Internal server error fetching resumes' })
  }
}

// ── Delete resume ───────────────────────────────────────────────────────────
export const deleteResume = async (req: AuthenticatedRequest, res: Response) => {
  if (!req.user) return res.status(401).json({ error: 'Unauthorized' })
  const id = req.params.id as string

  try {
    const resume = await prisma.resume.findUnique({ where: { id } })
    if (!resume) return res.status(404).json({ error: 'Resume not found' })
    if (resume.userId !== req.user.id) return res.status(403).json({ error: 'Forbidden' })

    // Remove the physical file if it exists locally
    if (resume.fileUrl.startsWith('/uploads/')) {
      const filePath = path.join(process.cwd(), resume.fileUrl)
      if (fs.existsSync(filePath)) fs.unlinkSync(filePath)
    }

    await prisma.resume.delete({ where: { id } })
    await calculateReadinessScore(req.user.id)
    res.json({ success: true })
  } catch (error) {
    res.status(500).json({ error: 'Internal server error deleting resume' })
  }
}

File: fastapi-backend/app/routers/resume.py
import os
import uuid
import json
import shutil
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File
from sqlmodel import Session, select

from app.database import get_session
from app.models import Resume, Profile, Skill
from app.auth import get_current_user
from app.routers.profile import calculate_readiness_score

router = APIRouter(prefix="/resume", tags=["resume"])

RESUMES_DIR = os.path.join(os.getcwd(), "uploads", "resumes")
os.makedirs(RESUMES_DIR, exist_ok=True)

def analyze_resume(user_id: str, filename: str, session: Session):
    profile = session.exec(select(Profile).where(Profile.userId == user_id)).first()
    
    user_skills = []
    if profile:
        skills = session.exec(select(Skill).where(Skill.profileId == profile.id)).all()
        user_skills = [s.name.lower() for s in skills]

    ats_keywords = ['javascript', 'typescript', 'python', 'react', 'node', 'sql', 'git', 'docker', 'ci/cd', 'system design']
    matched = [kw for kw in ats_keywords if kw in user_skills]
    missing = [kw for kw in ats_keywords if kw not in user_skills]

    base_score = 40
    skill_bonus = min(len(matched) * 6, 36)
    profile_bonus = 0
    if profile:
        profile_bonus += 8 if profile.bio else 0
        profile_bonus += 8 if profile.collegeName else 0
    profile_bonus += 8 if len(user_skills) >= 3 else 0
    
    ats_score = min(base_score + skill_bonus + profile_bonus, 100)

    keyword_report = [
        f'✅ Resume file "{filename}" successfully uploaded and parsed.',
        f'✅ {len(matched)} technical keywords matched: {", ".join(matched) if matched else "None yet — add skills in your portfolio."}',
        f'⚠️ ATS score is {ats_score}%. Add more skills and fill your bio to improve visibility.' if ats_score < 70 else f'✅ ATS score of {ats_score}% — excellent keyword optimization!',
        f'💡 Consider adding these keywords if you have them: "{", ".join(missing[:4])}"' if missing else '💡 Tip: Add your LinkedIn and GitHub URLs to your resume header for extra recruiter visibility.'
    ]

    return ats_score, keyword_report

@router.post("/upload", status_code=201)
async def upload_resume(file: UploadFile = File(...), current_user: dict = Depends(get_current_user), session: Session = Depends(get_session)):
    user_id = current_user.get("id")
    
    ext = os.path.splitext(file.filename)[1].lower()
    if ext not in [".pdf", ".docx", ".doc"]:
        raise HTTPException(status_code=400, detail="Only PDF and DOCX files are accepted")

    safe_name = "".join(c if c.isalnum() or c in "._-" else "_" for c in file.filename)
    filename = f"{user_id}_{int(datetime.utcnow().timestamp())}_{safe_name}"
    dest_path = os.path.join(RESUMES_DIR, filename)

    # Save physical file
    with open(dest_path, "wb") as buffer:
        shutil.copyfileobj(file.file, buffer)

    file_url = f"/uploads/resumes/{filename}"
    ats_score, report = analyze_resume(user_id, file.filename, session)

    new_resume = Resume(
        id=str(uuid.uuid4()),
        userId=user_id,
        fileName=file.filename,
        fileUrl=file_url,
        atsScore=ats_score,
        keywordReport=json.dumps(report),
        uploadedAt=datetime.utcnow()
    )
    session.add(new_resume)
    session.commit()
    session.refresh(new_resume)

    # Re-calc readiness score
    calculate_readiness_score(user_id, session)

    return new_resume

@router.get("/")
async def get_resumes(current_user: dict = Depends(get_current_user), session: Session = Depends(get_session)):
    user_id = current_user.get("id")
    resumes = session.exec(
        select(Resume).where(Resume.userId == user_id).order_by(Resume.uploadedAt.desc())
    ).all()
    return resumes

@router.delete("/{resume_id}")
async def delete_resume(resume_id: str, current_user: dict = Depends(get_current_user), session: Session = Depends(get_session)):
    user_id = current_user.get("id")
    resume = session.exec(select(Resume).where(Resume.id == resume_id)).first()
    if not resume:
        raise HTTPException(status_code=404, detail="Resume not found")

    if resume.userId != user_id:
        raise HTTPException(status_code=403, detail="Forbidden")

    # Delete physical file
    if resume.fileUrl.startswith("/uploads/"):
        filePath = os.path.join(os.getcwd(), resume.fileUrl.lstrip("/"))
        if os.path.exists(filePath):
            try:
                os.remove(filePath)
            except Exception:
                pass

    session.delete(resume)
    session.commit()
    
    # Re-calc readiness score
    calculate_readiness_score(user_id, session)

    return {"success": True}
